# Step 1
Create a movie recommendation system by processing the data from "tmdb_5000_credits.csv" and "tmdb_5000_movies.csv". The final output should be three pickle files: `movies.pkl`, `movies_dicti.pkl`, and `similarity.pkl`.

## Step 2

Load the `tmdb_5000_credits.csv` and `tmdb_5000_movies.csv` files into pandas DataFrames.


**Reasoning**:
Import pandas and load the two CSV files into dataframes, then display the head and info of both dataframes to inspect the data as instructed.



In [ ]:
import pandas as pd

credits_df = pd.read_csv('/content/tmdb_5000_credits.csv')
movies_df = pd.read_csv('/content/tmdb_5000_movies.csv')

print("Credits DataFrame Head:")
display(credits_df.head())
print("\nCredits DataFrame Info:")
credits_df.info()

print("\nMovies DataFrame Head:")
display(movies_df.head())
print("\nMovies DataFrame Info:")
movies_df.info()

Credits DataFrame Head:


movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...


Credits DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4803 non-null   int64 
 1   title     4803 non-null   object
 2   cast      4803 non-null   object
 3   crew      4803 non-null   object
dtypes: int64(1), object(3)
memory usage: 150.2+ KB

Movies DataFrame Head:


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124


Movies DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 1

## Step 3

Merge the two DataFrames on the 'title' column.


**Reasoning**:
Merge the two dataframes on the 'title' column as instructed and then display the head and info of the merged dataframe.



In [ ]:
merged_df = movies_df.merge(credits_df, on='title')

print("Merged DataFrame Head:")
display(merged_df.head())
print("\nMerged DataFrame Info:")
merged_df.info()

Merged DataFrame Head:


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...


Merged DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 1

## Step 4 Data preprocessing
Clean and preprocess the merged data, including handling missing values and selecting relevant features.


**Reasoning**:
Select relevant columns, handle missing values, and convert string representations of lists/dictionaries to actual lists/dictionaries.



In [ ]:
import ast

# Select relevant columns
merged_df = merged_df[['movie_id', 'title', 'overview', 'genres', 'keywords', 'cast', 'crew']]

# Handle missing values in 'overview'
merged_df['overview'] = merged_df['overview'].fillna('')

# Drop rows with missing values in 'genres', 'keywords', 'cast', 'crew' if any
merged_df.dropna(subset=['genres', 'keywords', 'cast', 'crew'], inplace=True)

# Convert string representations of lists/dictionaries to actual lists/dictionaries
features = ['genres', 'keywords', 'cast', 'crew']
for feature in features:
    merged_df[feature] = merged_df[feature].apply(ast.literal_eval)

display(merged_df.head())
merged_df.info()

/tmp/ipython-input-2516643369.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df['overview'] = merged_df['overview'].fillna('')
/tmp/ipython-input-2516643369.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df.dropna(subset=['genres', 'keywords', 'cast', 'crew'], inplace=True)
/tmp/ipython-input-2516643369.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres  \
0  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3  [{'id': 28, 'name': 'Action'}, {'id': 80, 'nam...   
4  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                            keywords  \
0  [{'id': 1463, 'name': 'culture clash'}, {'id':...   
1  [{'id': 270, 'name': 'ocean'}, {'id': 726, 'na...   
2  [{'id': 470, 'name': 'spy'}, {'id': 818, 'name...   
3  [{'id': 849, 'name': 'dc comics'}, {'id': 853,...   
4  [{'id': 818, 'name': 'based on novel'}, {'id':...   

                                                cast  \
0  [{'cast_id': 242, 'character': 'Jake Sully', '...   
1  [{'cast_id': 4, 'character': 'Captain Jack Spa...   
2  [{'cast_id': 1, 'character': 'James Bond', 'cr...   
3  [{'cast_id': 2, 'character': 'Bruce Wayne / Ba...   
4  [{'cast_id': 5, 'character': 'John Carter', 'c...   

                                                crew  
0  [{'credit_id': '52fe48009251416c750aca23', 'de...  
1  [{'credit_id': '52fe4232c3a36847f800b579', 'de...  
2  [{'credit_id': '54805967c3a36829b5002c41', 'de...  
3  [{'credit_id': '52fe4781c3a36847f81398c3', 'de...  
4  [{'credit_id': '52fe479ac3a36847f813eaa3', 'de...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4809 non-null   int64 
 1   title     4809 non-null   object
 2   overview  4809 non-null   object
 3   genres    4809 non-null   object
 4   keywords  4809 non-null   object
 5   cast      4809 non-null   object
 6   crew      4809 non-null   object
dtypes: int64(1), object(6)
memory usage: 263.1+ KB


## Step 5 Feature extraction


Extract features like genres, keywords, cast, and crew.


**Reasoning**:
Define and apply functions to extract features from the 'genres', 'keywords', 'cast', and 'crew' columns, then preprocess these features and concatenate them into a 'tags' column. Finally, display the relevant parts of the dataframe to verify the results.



In [ ]:
def get_top_n_names(list_of_dicts, n=3):
    """Extracts the top N names from a list of dictionaries."""
    if not isinstance(list_of_dicts, list):
        return []
    names = [item['name'] for item in list_of_dicts if isinstance(item, dict) and 'name' in item]
    return names[:min(len(names), n)]

def get_director_name(list_of_dicts):
    """Extracts the director's name from a list of dictionaries in the 'crew' column."""
    if not isinstance(list_of_dicts, list):
        return ''
    for item in list_of_dicts:
        if isinstance(item, dict) and item.get('job') == 'Director':
            return item.get('name', '')
    return ''

# Apply the functions to extract features
merged_df['genres'] = merged_df['genres'].apply(get_top_n_names)
merged_df['keywords'] = merged_df['keywords'].apply(get_top_n_names)
merged_df['cast'] = merged_df['cast'].apply(get_top_n_names)
merged_df['director'] = merged_df['crew'].apply(get_director_name)

# Preprocess the extracted features
def preprocess_text(text):
    """Converts text to lowercase and removes spaces."""
    if isinstance(text, list):
        return [str(i).lower().replace(" ", "") for i in text]
    elif isinstance(text, str):
        return text.lower().replace(" ", "")
    else:
        return ""

merged_df['genres'] = merged_df['genres'].apply(preprocess_text)
merged_df['keywords'] = merged_df['keywords'].apply(preprocess_text)
merged_df['cast'] = merged_df['cast'].apply(preprocess_text)
merged_df['director'] = merged_df['director'].apply(preprocess_text)

# Concatenate features into 'tags' column
merged_df['tags'] = merged_df['overview'] + ' ' + merged_df['genres'].apply(lambda x: ' '.join(x)) + ' ' + \
                    merged_df['keywords'].apply(lambda x: ' '.join(x)) + ' ' + \
                    merged_df['cast'].apply(lambda x: ' '.join(x)) + ' ' + merged_df['director']

# Display results
display(merged_df[['title', 'genres', 'keywords', 'cast', 'director', 'tags']].head())
merged_df.info()

title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                genres                             keywords  \
0         [action, adventure, fantasy]     [cultureclash, future, spacewar]   
1         [adventure, fantasy, action]     [ocean, drugabuse, exoticisland]   
2           [action, adventure, crime]     [spy, basedonnovel, secretagent]   
3               [action, crime, drama]  [dccomics, crimefighter, terrorist]   
4  [action, adventure, sciencefiction]      [basedonnovel, mars, medallion]   

                                            cast          director  \
0  [samworthington, zoesaldana, sigourneyweaver]      jamescameron   
1     [johnnydepp, orlandobloom, keiraknightley]     goreverbinski   
2      [danielcraig, christophwaltz, léaseydoux]         sammendes   
3      [christianbale, michaelcaine, garyoldman]  christophernolan   
4    [taylorkitsch, lynncollins, samanthamorton]     andrewstanton   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4809 non-null   int64 
 1   title     4809 non-null   object
 2   overview  4809 non-null   object
 3   genres    4809 non-null   object
 4   keywords  4809 non-null   object
 5   cast      4809 non-null   object
 6   crew      4809 non-null   object
 7   director  4809 non-null   object
 8   tags      4809 non-null   object
dtypes: int64(1), object(8)
memory usage: 338.3+ KB


## Step 6


Compute a similarity matrix based on the extracted features.


**Reasoning**:
Compute the cosine similarity matrix based on the 'tags' column.



In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# Instantiate TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_features=5000)

# Fit and transform the 'tags' column
tfidf_matrix = tfidf_vectorizer.fit_transform(merged_df['tags'])

# Compute cosine similarity
similarity = cosine_similarity(tfidf_matrix)

print("Shape of TF-IDF matrix:", tfidf_matrix.shape)
print("Shape of similarity matrix:", similarity.shape)

Shape of TF-IDF matrix: (4809, 5000)
Shape of similarity matrix: (4809, 4809)


## Step 7

Create a dictionary of movies for easy lookup.


**Reasoning**:
Create a dictionary to map movie titles to their indices in the merged_df DataFrame and print the first few items.



In [ ]:
movies_dicti = merged_df.reset_index().set_index('title')['index'].to_dict()

# Print the first few key-value pairs
print("First 5 items in movies_dicti:")
for i, (key, value) in enumerate(movies_dicti.items()):
    if i >= 5:
        break
    print(f"{key}: {value}")

First 5 items in movies_dicti:
Avatar: 0
Pirates of the Caribbean: At World's End: 1
Spectre: 2
The Dark Knight Rises: 3
John Carter: 4


## Step 8

Save the processed movie data, the movie dictionary, and the similarity matrix as pickle files (`movies.pkl`, `movies_dicti.pkl`, `similarity.pkl`).


**Reasoning**:
Save the processed dataframes and the similarity matrix to pickle files as instructed.



In [ ]:
import pickle

# Save merged_df to movies.pkl
with open('movies.pkl', 'wb') as f:
    pickle.dump(merged_df, f)

# Save movies_dicti to movies_dicti.pkl
with open('movies_dicti.pkl', 'wb') as f:
    pickle.dump(movies_dicti, f)

# Save similarity to similarity.pkl
with open('similarity.pkl', 'wb') as f:
    pickle.dump(similarity, f)

print("Files saved successfully: movies.pkl, movies_dicti.pkl, similarity.pkl")

Files saved successfully: movies.pkl, movies_dicti.pkl, similarity.pkl


## Summary:

### Data Analysis Key Findings

*   Two datasets, `tmdb_5000_credits.csv` and `tmdb_5000_movies.csv`, were loaded and merged based on the 'title' column, resulting in a combined DataFrame with 4809 entries.
*   Relevant columns for recommendation ('movie\_id', 'title', 'overview', 'genres', 'keywords', 'cast', 'crew') were selected, and missing values in 'overview' were handled by filling them with empty strings.
*   String representations of lists/dictionaries in 'genres', 'keywords', 'cast', and 'crew' were converted to actual Python lists/dictionaries.
*   Functions were applied to extract the top 3 genres, keywords, and cast members, as well as the director's name.
*   The extracted features were preprocessed by converting them to lowercase and removing spaces.
*   A 'tags' column was created by concatenating the 'overview' and the processed 'genres', 'keywords', 'cast', and 'director' information.
*   A TF-IDF matrix with a shape of (4809, 5000) was generated from the 'tags' column using `TfidfVectorizer`.
*   A cosine similarity matrix with a shape of (4809, 4809) was computed based on the TF-IDF matrix.
*   A dictionary named `movies_dicti` was created, mapping movie titles to their corresponding indices in the processed DataFrame.
*   The processed DataFrame (`merged_df`), the movie dictionary (`movies_dicti`), and the similarity matrix (`similarity`) were successfully saved as `movies.pkl`, `movies_dicti.pkl`, and `similarity.pkl` respectively.

### Insights or Next Steps

*   The generated pickle files (`movies.pkl`, `movies_dicti.pkl`, and `similarity.pkl`) are now available for building a movie recommendation function that can take a movie title as input and return a list of similar movies based on the calculated cosine similarity.
*   Further evaluation of the recommendation system's performance could be conducted using metrics like precision, recall, or diversity of recommendations.


# Step 1
Create a movie recommendation system using Streamlit, based on the provided data files: "movies.pkl", "movies_dicti.pkl", and "similarity.pkl". The system should allow users to select a movie from a dropdown and display recommended movies.

## Step 2 Install streamlit

Install the Streamlit library.


**Reasoning**:
The subtask is to install the Streamlit library. I will use the `pip install` command within a code cell to install it.



In [ ]:
%pip install streamlit

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.1/10.1 MB 64.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 107.4 MB/s eta 0:00:00


## Step 3

Load the `movies.pkl`, `movies_dicti.pkl`, and `similarity.pkl` files.


**Reasoning**:
Load the pickle files containing the movie data, dictionary, and similarity matrix.



In [ ]:
import pickle

with open('movies.pkl', 'rb') as f:
    movies = pickle.load(f)

with open('movies_dicti.pkl', 'rb') as f:
    movies_dicti = pickle.load(f)

with open('similarity.pkl', 'rb') as f:
    similarity = pickle.load(f)

print("Pickle files loaded successfully.")

Pickle files loaded successfully.


## Step 4

Define a function that takes a movie title as input and returns a list of recommended movies using the loaded data.


**Reasoning**:
Define the `recommend` function to get movie recommendations based on similarity scores.



In [ ]:
def recommend(movie):
    """Recommends movies similar to the input movie based on cosine similarity."""
    if movie not in movies_dicti:
        return "Movie not found."

    movie_index = movies_dicti[movie]
    distances = similarity[movie_index]
    movies_list = sorted(list(enumerate(distances)), reverse=True, key=lambda x: x[1])[1:6]

    recommended_movies = []
    for i in movies_list:
        recommended_movies.append(movies.iloc[i[0]].title)
    return recommended_movies

# Test the function
print("Recommendations for Avatar:")
print(recommend('Avatar'))
print("\nRecommendations for a non-existent movie:")
print(recommend('NonExistentMovie'))

Recommendations for Avatar:
['Apollo 18', 'The Book of Life', 'The American', 'Tears of the Sun', 'Falcon Rising']

Recommendations for a non-existent movie:
Movie not found.


## Step 5

Create a Streamlit application with a dropdown menu to select a movie and a button to get recommendations.


**Reasoning**:
Create a Streamlit application with a dropdown menu to select a movie and a button to get recommendations as instructed.



In [ ]:
import streamlit as st

st.title('Movie Recommender System')

selected_movie_name = st.selectbox(
    'Select a movie:',
    (movies_dicti.keys()))

if st.button('Get Recommendation'):
    recommendations = recommend(selected_movie_name)
    st.write('Recommended Movies:')
    for movie in recommendations:
        st.write(movie)

2025-10-21 03:36:40.088 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-10-21 03:36:40.432 
  command:

    streamlit run /usr/local/lib/python3.12/dist-packages/colab_kernel_launcher.py [ARGUMENTS]
2025-10-21 03:36:40.434 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-10-21 03:36:40.437 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-10-21 03:36:40.439 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-10-21 03:36:40.442 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-10-21 03:36:40.447 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-10-21 03:36:40.454 Thread 'MainThread': mi

## Run the streamlit app

Provide instructions on how to run the Streamlit application locally.


## Summary:

### Data Analysis Key Findings

*   The `streamlit` library was successfully installed to build the web application.
*   The necessary data files (`movies.pkl`, `movies_dicti.pkl`, and `similarity.pkl`) were successfully loaded using the `pickle` library.
*   A Python function `recommend` was defined and tested, which takes a movie title as input and returns a list of 5 recommended movies based on pre-calculated cosine similarity scores. The function correctly handles cases where the input movie is not found.
*   A basic Streamlit application structure was created with a title, a dropdown menu for selecting a movie, and a button to trigger the recommendation process.

### Insights or Next Steps

*   The created Streamlit application provides a functional user interface for the movie recommendation system, allowing users to easily select a movie and view recommendations.
*   The next step is to integrate the `recommend` function with the Streamlit application so that clicking the "Get Recommendation" button displays the actual recommendations generated by the function.
